runner/entrypoint.sh

```diff
@@ -44,63 +64,85 @@ if [ -z "${RUNNER_REPO}" ] && [ -n "${RUNNER_GROUP}" ];then
-  # "agentName": "THE_RUNNER_POD_NAME",
-  # "poolId": 1,
-  # "poolName": "Default",
-  # "serverUrl": "https://pipelines.actions.githubusercontent.com/SOME_RANDOM_ID",
-  # "gitHubUrl": "https://github.com/USER/REPO",
-  # "workFolder": "/some/work/dir" #=> corresponds to Runner.Spec.WorkDir
-  # }
-  #
-  # Especially `agentId` is important, as other than listing all the runners in the repo,
-  # this is the only change we could get the exact runnner ID which can be useful for further
-  # GitHub API call like the below. Note that 171 is the agentId seen above.
-  #   curl \
-  #     -H "Accept: application/vnd.github.v3+json" \
-  #     -H "Authorization: bearer ${GITHUB_TOKEN}"
-  #     https://api.github.com/repos/USER/REPO/actions/runners/171
+# if this is not a testing environment
+if [ -z "${UNITTEST:-}" ]; then
+  sudo chown -R runner:docker ${RUNNER_HOME}
+  mv /runnertmp/* ${RUNNER_HOME}/
 fi
 
+cd ${RUNNER_HOME}
+# past that point, it's all relative pathes from /runner
+
+retries_left=10
+while [[ ${retries_left} -gt 0 ]]; do
+  log "Configuring the runner."
+  ./config.sh --unattended --replace \
+    --name "${RUNNER_NAME}" \
+    --url "${GITHUB_URL}${ATTACH}" \
+    --token "${RUNNER_TOKEN}" \
+    --runnergroup "${RUNNER_GROUPS}" \
+    --labels "${RUNNER_LABELS}" \
+    --work "${RUNNER_WORKDIR}"
+
+  if [ -f .runner ]; then
+    success "Runner successfully configured."
+    break
+  fi
+
+  error "Configuration failed. Retrying"
+  retries_left=$((retries_left - 1))
+  sleep 1
+done
+
+if [ ! -f .runner ]; then
+  # we couldn't configure and register the runner; no point continuing
+  error "Configuration failed!"
+  exit 2
+fi
+
+cat .runner
+# Note: the `.runner` file's content should be something like the below:
+#
+# $ cat /runner/.runner
+# {
+# "agentId": 117, #=> corresponds to the ID of the runner
+# "agentName": "THE_RUNNER_POD_NAME",
+# "poolId": 1,
+# "poolName": "Default",
+# "serverUrl": "https://pipelines.actions.githubusercontent.com/SOME_RANDOM_ID",
+# "gitHubUrl": "https://github.com/USER/REPO",
+# "workFolder": "/some/work/dir" #=> corresponds to Runner.Spec.WorkDir
+# }
+#
+# Especially `agentId` is important, as other than listing all the runners in the repo,
+# this is the only change we could get the exact runnner ID which can be useful for further
+# GitHub API call like the below. Note that 171 is the agentId seen above.
+#   curl \
+#     -H "Accept: application/vnd.github.v3+json" \
+#     -H "Authorization: bearer ${GITHUB_TOKEN}"
+#     https://api.github.com/repos/USER/REPO/actions/runners/171
+
 if [ -n "${RUNNER_REGISTRATION_ONLY}" ]; then
-  echo
-  echo "This runner is configured to be registration-only. Exiting without starting the runner service..."
+  success "This runner is configured to be registration-only. Exiting without starting the runner service..."
   exit 0
 fi
 
-mkdir ./externals
-# Hack due to the DinD volumes
-mv ./externalstmp/* ./externals/
+if [ -z "${UNITTEST:-}" ]; then
```

Comment:
same as line 73, This is hard to mock, so had to add a conditional here.

Reply:
Ah, good catch!

I would rather make this a standard feature behind a variable like `RUNNER_CREATE_EXTERNALS_DIR_FROM` so that L141 to L144 would look like:

```
RUNNER_CREATE_EXTERNALS_DIR_FROM=${RUNNER_CREATE_EXTERNALS_DIR_FROM:-./externalstmp}
if [ ! -z "${RUNNER_CREATE_EXTERNALS_DIR_FROM}" ]; then
  mkdir ./externals
  # Hack due to the DinD volumes
  mv ${RUNNER_CREATE_EXTERNALS_DIR_FROM}/* ./externals/
fi
```

Similarly, `runsvc.sh` and `RunnerService.js` patching can be behind a variable so that it looks like:

```
RUNNER_PATCH_SERVICE_FILES=${RUNNER_PATCH_SERVICE_FILES:-true}
if [ ! -z "${RUNNER_PATCH_SERVICE_FILES}" ]; then
  for f in runsvc.sh RunnerService.js; do
    diff {bin,patched}/${f} || :
    sudo mv bin/${f}{,.bak}
    sudo mv {patched,bin}/${f}
  done
fi
```

`RUNNER_CREATE_EXTERNALS_DIR_FROM` and `RUNNER_PATCH_SERVICE_FILES` would be unnecessary in unit tests so I expect you to unset those variables in tests.